### PR TITLE
Associated types in Contract

### PR DESCRIPTION
**Pull request by christos-h**
# Motivation

Our existing Contracts receive and return vectors of bytes. This results application writers to have to constantly implement serde themselves.

# Solution

Introduce associated types in contracts with the appropriate trait bounds.

# Limitations

There is still an open questions regarding error management when converting to and from wit types, this issue is intended to tackle this and we can merge this PR regardless for now: #743.

### CODE REVIEW

File: linera-examples/fungible2/src/contract.rs [hunk 1]
```diff
@@ -25,14 +25,14 @@ linera_sdk::contract!(FungibleToken<ViewStorageContext>);
...
 
     async fn initialize(
         &mut self,
         context: &OperationContext,
-        argument: &[u8],
+        argument: Self::InitializationArguments,
     ) -> Result<ExecutionResult, Self::Error> {
-        let mut accounts: BTreeMap<AccountOwner, Amount> =
-            bcs::from_bytes(argument).map_err(Error::InvalidInitialState)?;
+        let mut accounts = argument;
```
Comment (ma2bd) [inline linera-examples/fungible2/src/contract.rs:35]: Of course this will be simplified.

File: linera-examples/fungible2/src/contract.rs [hunk 2]
```diff
@@ -262,3 +262,9 @@ pub enum Error {
     #[error("The requested transfer is not correctly authenticated.")]
     IncorrectAuthentication,
 }
+
+impl From<bcs::Error> for Error {
```
Comment (ma2bd) [inline linera-examples/fungible2/src/contract.rs:266]: You shouldn't need to write that (use `#[from]`).

File: linera-sdk/src/lib.rs [hunk 3]
```diff
@@ -93,9 +95,25 @@ pub trait Contract: Sized {
...
         &mut self,
         context: &OperationContext,
-        argument: &[u8],
+        argument: Self::InitializationArguments,
     ) -> Result<ExecutionResult, Self::Error>;
 
+    /// Initializes the application on the chain that created it.
+    ///
+    /// Performs serialization from `&[u8]` into [`Contract::InitializationArguments`] and delegates
+    /// to [`Contract::InitializationArguments`].
```
Comment (jvff) [inline linera-sdk/src/lib.rs:104]: Nit: I'd reword this documentation a bit, into something like:

```suggestion
    /// Deserializes the `argument` into [`Contract::InitializationArguments`] and calls
    /// [`Contract::initialize`].
```

**Review by jvff — COMMENTED** [review]
Looks great! :tada:

File: linera-sdk/src/lib.rs [hunk 4]
```diff
@@ -75,9 +75,11 @@ pub struct ViewStateStorage<A>(std::marker::PhantomData<A>);
 #[async_trait]
 pub trait Contract: Sized {
     /// Message reports for application execution errors.
-    type Error: Error;
+    type Error: Error + From<bcs::Error>;
     /// The desired storage backend to use to store the application's state.
     type Storage;
+    /// Initialization Arguments.
+    type InitializationArguments: FromBcsBytes + Send;
```
Comment (ma2bd) [inline linera-sdk/src/lib.rs:82]: The convenient trait `FromBcsBytes` becomes unnecessary because users won't see it. (In other words, we might just use `serde::Deserialize` here.)

File: linera-sdk/src/lib.rs [hunk 5]
```diff
@@ -93,9 +95,25 @@ pub trait Contract: Sized {
...
-        argument: &[u8],
+        argument: Self::InitializationArguments,
     ) -> Result<ExecutionResult, Self::Error>;
 
+    /// Initializes the application on the chain that created it.
+    ///
+    /// Performs serialization from `&[u8]` into [`Contract::InitializationArguments`] and delegates
+    /// to [`Contract::InitializationArguments`].
+    ///
+    /// Do not override this method unless you really know what you're doing.
```
Comment (ma2bd) [inline linera-sdk/src/lib.rs:106]: I don't think this should be allowed -> this code belongs to the SDK.

Comment (afck) [reply linera-sdk/src/lib.rs:106]: Instead of adding it to the trait, could the `contract!` macro just generate this method?

File: linera-examples/fungible2/src/contract.rs (hunk 1 above)
Comment (christos-h) [reply linera-examples/fungible2/src/contract.rs:35]: This is just essentially aliasing the argument.

File: linera-sdk/src/lib.rs (hunk 5 above)
Comment (christos-h) [reply linera-sdk/src/lib.rs:106]: @ma2bd great idea.

File: linera-examples/fungible2/src/contract.rs (hunk 1 above)
Comment (ma2bd) [reply linera-examples/fungible2/src/contract.rs:35]: I know but in this case one can use `mut argument` in the function header

File: linera-sdk/src/lib.rs [hunk 6]
```diff
@@ -79,9 +79,23 @@ pub struct ViewStateStorage<A>(std::marker::PhantomData<A>);
...
+    /// Initialization Arguments.
+    type InitializationArguments: DeserializeOwned + Send;
+    /// Execute Operation Arguments.
+    type Operation: DeserializeOwned + Send;
+    /// Application Call Arguments.
+    type ApplicationCallArguments: DeserializeOwned + Serialize + Send;
+    /// The Application Effect.
+    type Effect: DeserializeOwned + Serialize + Send + std::fmt::Debug;
+    /// A Session Call.
+    type SessionCall: DeserializeOwned + Send;
```
Comment (ma2bd) [inline linera-sdk/src/lib.rs:94]: Would `SessionCallArguments` be more logical?

File: linera-sdk/src/lib.rs [hunk 7]
```diff
@@ -79,9 +79,23 @@ pub struct ViewStateStorage<A>(std::marker::PhantomData<A>);
...
+    /// The response type of an application call.
+    type Response: Serialize + Send;
+    /// The type for the contract's sessions.
+    type Session: DeserializeOwned + Serialize + Send;
```
Comment (ma2bd) [inline linera-sdk/src/lib.rs:98]: `SessionState` ?

File: linera-sdk/src/lib.rs [hunk 8]
```diff
@@ -265,71 +279,84 @@ pub struct QueryContext {
...
-#[derive(Debug, Default, Deserialize, Serialize)]
+#[derive(Debug, Deserialize, Serialize)]
 #[cfg_attr(any(test, feature = "test"), derive(Eq, PartialEq))]
-pub struct ApplicationCallResult {
+pub struct ApplicationCallResult<Effect, Value, SessionType> {
     /// The return value.
-    #[debug(with = "linera_base::hex_debug")]
-    pub value: Vec<u8>,
+    // #[debug(with = "linera_base::hex_debug")]
+    pub value: Option<Value>,
```
Comment (ma2bd) [inline linera-sdk/src/lib.rs:328]: I think we should use `Value` here and let the authors of each application decide if they want `Self::ApplicationCallReturnValue` to implement `Default` in order to benefit from the `impl Default` below.

File: linera-sdk/src/lib.rs [hunk 9]
```diff
@@ -265,71 +279,84 @@ pub struct QueryContext {
...
+            create_sessions: vec![],
+        }
+    }
 }
 
 /// Syscall to request creating a new session.
 #[derive(Debug, Default, Deserialize, Serialize)]
 #[cfg_attr(any(test, feature = "test"), derive(Eq, PartialEq))]
-pub struct Session {
+pub struct Session<T> {
```
Comment (ma2bd) [inline linera-sdk/src/lib.rs:348]: Following your observation offline, this could be just `T = Self::SessionState` I guess.

File: linera-sdk/src/lib.rs [hunk 10]
```diff
@@ -265,71 +279,84 @@ pub struct QueryContext {
...
 
 /// The result of calling into a session.
 #[derive(Default, Deserialize, Serialize)]
-pub struct SessionCallResult {
+pub struct SessionCallResult<Effect, Value, Session> {
     /// The application result.
-    pub inner: ApplicationCallResult,
+    pub inner: ApplicationCallResult<Effect, Value, Session>,
     /// If `call_session` was called, this tells the system to clean up the session.
     pub data: Option<Vec<u8>>,
```
Comment (ma2bd) [inline linera-sdk/src/lib.rs:361]: I believe this `Vec<u8>` is meant to be the session state (see above). `data` could be called `new_state`. (`None` means discard the object)

File: linera-sdk/src/lib.rs [hunk 11]
```diff
@@ -79,9 +79,23 @@ pub struct ViewStateStorage<A>(std::marker::PhantomData<A>);
...
+    /// Execute Operation Arguments.
+    type Operation: DeserializeOwned + Send;
+    /// Application Call Arguments.
+    type ApplicationCallArguments: DeserializeOwned + Serialize + Send;
+    /// The Application Effect.
+    type Effect: DeserializeOwned + Serialize + Send + std::fmt::Debug;
+    /// A Session Call.
+    type SessionCall: DeserializeOwned + Send;
+    /// The response type of an application call.
+    type Response: Serialize + Send;
```
Comment (ma2bd) [inline linera-sdk/src/lib.rs:96]: Seems like this should be called `ApplicationCallReturnValue` then (and moved above).

Comment (ma2bd) [reply linera-sdk/src/lib.rs:96]: Oh I see. You wanted to re-use the same type as sessions. But I don't see why they would be the same type in general. Looks like we need `SessionCallReturnValue`.

File: linera-sdk/src/contract/mod.rs [hunk 12]
```diff
@@ -41,18 +41,18 @@ macro_rules! contract {
                 &mut self,
                 authenticated: bool,
                 application: $crate::base::ApplicationId,
-                argument: &[u8],
+                argument: &impl serde::Serialize,
```
Comment (ma2bd) [inline linera-sdk/src/contract/mod.rs:44]: We can do better than this now

File: linera-sdk/src/contract/mod.rs [hunk 13]
```diff
@@ -41,18 +41,18 @@ macro_rules! contract {
                 &mut self,
```
Comment (ma2bd) [inline linera-sdk/src/contract/mod.rs:41]: @jvff Shouldn't this have been a method with a default implementation in the `Contract` trait itself (or an extension)?

File: linera-sdk/src/contract/conversions_to_wit.rs [hunk 14]
```diff
@@ -78,46 +79,67 @@ impl From<log::Level> for system::LogLevel {
...
+    fn from(result: ApplicationCallResult<Effect, Value, Session>) -> Self {
         let create_sessions = result
             .create_sessions
             .into_iter()
             .map(wit_types::Session::from)
             .collect();
 
+        let value = result
+            .value
+            .map(|v| bcs::to_bytes(&v).expect("TODO"))
```
Comment (ma2bd) [inline linera-sdk/src/contract/conversions_to_wit.rs:98]: What's the plan for the serialization error? `TryFrom` or panic?

File: examples/fungible/src/contract.rs [hunk 15]
```diff
@@ -146,28 +145,29 @@ impl FungibleToken<ViewStorageContext> {
...
-            value: balance_bytes,
-            ..Default::default()
+            value: Some(session_data),
+            execution_result: ExecutionResult::default(),
+            create_sessions: vec![],
         };
         let session_call_result = SessionCallResult {
             inner: application_call_result,
-            data: Some(session_data),
+            data: Some(bcs::to_bytes(&session_data).expect("TODO")),
```
Comment (ma2bd) [inline examples/fungible/src/contract.rs:159]: see other comment

File: linera-sdk/src/contract/conversions_to_wit.rs (hunk 14 above)
Comment (christos-h) [reply linera-sdk/src/contract/conversions_to_wit.rs:98]: If `bcs` serde is symmetric, i.e. if serialization for a given struct doesn't fail it is guaranteed that deserialization won't fail, and vice versa, we could panic since our runtime is soleley responsible for performing the serde?

@jvff any thoughts on this?

File: linera-sdk/src/lib.rs (hunk 9 above)
Comment (christos-h) [reply linera-sdk/src/lib.rs:348]: I'm not sure I understand this point? Are you referring to removing `kind` and therefore this entire struct becomes redundant?

File: linera-sdk/src/contract/conversions_to_wit.rs (hunk 14 above)
Comment (jvff) [reply linera-sdk/src/contract/conversions_to_wit.rs:99]: I think we should probably list the BCS serialization error scenarios, just to be sure. Can an app developer reach any of them? If that's the case, `panic`s will block the application, which could be pretty bad.

If we can guarantee that there are no failure scenarios that an app developer can reach, I agree that we should panic.

I can start looking into this :+1:

File: linera-sdk/src/contract/mod.rs (hunk 13 above)
Comment (jvff) [reply linera-sdk/src/contract/mod.rs:41]: Good point. I can try that :+1:

File: linera-sdk/src/lib.rs (hunk 11 above)
Comment (christos-h) [reply linera-sdk/src/lib.rs:97]: Well it's wrapping an `ApplicationCallResult` right? I assumed they were the same.
